util/unitylauncher_util.go

```
@@ -0,0 +1,183 @@
+	defer k.Close()
+
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	info, err := k.Stat()
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	keyNames, err := k.ReadValueNames(int(info.ValueCount))
+	for _, keyName := range keyNames {
+		if strings.HasPrefix(keyName, recentlyUsedProjectKey) {
+			val, _, err := k.GetBinaryValue(keyName)
+			if err != nil {
+				continue
+			}
+
+			projects = append(projects, &PathAndVersion{Path: strings.Trim(string(val), "\x00")})
+		}
+	}
+
+	return
+}
+
+func LocateInstalledUnityVersions(unityInstallsPath string) (installations UnityInstallations) {
+	installations = UnityInstallations{Versions: make([]UnityVersion, 0), Installations: make(map[UnityVersion]PathAndVersion)}
+
+	fileInfos, err := ioutil.ReadDir(unityInstallsPath)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	re := regexp.MustCompile(VersionRegexString)
+
+	for _, fileInfo := range fileInfos {
+		if fileInfo.IsDir() {
+
+			versionName := fileInfo.Name()
+			version := ExtractUnityVersion(versionName, re)
+
+			unityExecPath := filepath.Join(unityInstallsPath, versionName, unityLaunchPath)
+			stat, err := os.Stat(unityExecPath)
+			fmt.Println(stat)
+
+			if !os.IsNotExist(err) {
+				installations.AddInstallation(PathAndVersion{Version: version, Path: unityExecPath})
+			}
+		}
+	}
+
+	return
+}
+
+var VersionRegexString = `((?P<Series>\d+)\.(?P<Major>\d+)\.(?P<Minor>\d+)\w(?P<Patch>\d+))`
+var versionRegex = regexp.MustCompile(`m_EditorVersion: ` + VersionRegexString)
+var versionRegexGroupNames = versionRegex.SubexpNames()
+
+func ExtractUnityVersion(input string, re *regexp.Regexp) (version UnityVersion) {
+	match := re.FindStringSubmatch(input)
+	result := make(map[string]string)
+
+	if len(match) < 4 {
+		fmt.Printf("Could not parse v string %s", input)
+		version = UnityVersion{}
+		return
+	}
+
+	for i, name := range versionRegexGroupNames {
+		if i > 0 {
+			result[name] = match[i]
+		}
+	}
+
+	series, _ := strconv.ParseUint(result["Series"], 0, 64)
```

Don't ignore errors! Never, unless formally proven that they can never, ever, ever, ever happen.